### DISCUSSION TITLE
What is the best testing framework to use with instructor?

### DISCUSSION THREAD

**Discussion by drobertson-dev** [567-labs/instructor#389] [Q&A]
Hello everyone, I was wondering what other people use as a testing framework for prompts etc. when you use instructor?
I am finding that most are either way to complicated, don't support function/tool calls or if they do is a second class citizen.  Example, promptfoo seemed promising but, I don't want to build my pydantic models, then create separate json schema that has to go inside of yaml, its messy and defeats the purpose of a test.  It seems like everyone forgot about function calls, yet what you can do with them is amazing.

**Comment by jxnl** [comment]
pytest

**Reply by drobertson-dev** [reply]
I suspected that might be the answer, my apologies for not explaining clearly. To test a prompt for a specific model in for example PromptFoo, I need the function schema in yaml or json in yaml. However, since I'm using Pydantic models to generate the schema via Instructor, there's a mismatch if I dump it from the model, plus this would be error-prone and to manually make a change from one to another isn't going to work. With 20-30 models, this becomes a challenge.

Although I'm aware of using pytest for unit tests, I'm curious about your recommendation for testing prompts in a complex application using Instructor.  If an application lifecycle uses a bunch of models, this get's tedious very quickly.

**Reply by jxnl** [reply]
why is a function that uses a prompt more special than python code as a unit or integration test?

**Reply by drobertson-dev** [reply]
Maybe it's my use case, models get chosen dynamically through the lifecycle of the application.  It's one of those thing you would need to see it. What I just said literally explains nothing. lol. I can't publicly put the application on here, but I could privately show you.  You would see what I mean if you could see it. It's pretty awesome what the structured data mindset allows you to acheive with LLMs.

**Reply by jxnl** [reply]
why not write unit tests for selection, and then unit test for extraction,

**Reply by drobertson-dev** [reply]
This is interesting.  I'm gonna really think about this, I love the simplicity of Instructor, in a lot of ways it's quite brilliant. Function calling in the LLM context throws people off, but once you get it, it's like a whole new world of possibilities opens up.  I think you're probably right, and I need to think of this in a different way. 

 Maybe this would help explain it better, i'm not really doing any extraction, but rather using the LLM via multiple concurrent "function or tool calls" to dynamically build a larger schema which then gets used in a different part of the application. So I'm not really using it to extract, but more to build out something larger from a series of smaller function calls.  The aim was to take a larger problem, break it down into more simple singular problems, which would then in turn solve the larger more difficult problem.

**Reply by jxnl** [reply]
Makes sense. 

I would challenge you to think harder about how to break down the code itself to be more testable.

**Reply by drobertson-dev** [reply]
I've broken it down some, but not enough. The example in the docs of the cache function was a bit of a lightbulb moment for code reuse.  I'm normally working with JS/TS and I don't really ever use classes.  I'm going to work on this and see if I can get it broken down better.  Thank you for the input, I appreciate it greatly.

**Reply by jxnl** [reply]
There's a JavaScript version actually! 

https://instructor-ai.github.io/instructor-js/

**Reply by drobertson-dev** [reply]
Nice!  That's going to be a challenge and a half to replicate the python version.  I have to say, I really like Pydantic a lot more than zod. I mean zod is pretty awesome, but I feel that Pydantic is more intuitive and more fine grained for sure, then add that to FastApi, I may never use server routes again. LOL.  

Thank You again for you help, i was able to achieve everything I needed to with pytest, I got some good ideas from the examples and boom, no need for some extra dependency.  I have a nice setup that allows me to test the open source along with the openai models if I want, which is nice.  It works great.  I'm so glad I saw your talk back in October, it really allowed me to think different about how to use AI, and because of that I was given a chance at my dream job, and now it will hopefully become a permanent position.
